syntax = "proto3";

package framework.protos;

import "framework/protos/node.proto";

enum Code {
  OK = 0;
  ERROR = 1;
}

enum MessageType {
  PLAIN = 0;
  CREATE_JOB = 1;
  QUERY_JOB = 2;
  FINISH_TASK = 3;
  START_TASK = 4;
  HEARTBEAT = 5;
}

message Message{
  Node node = 1;
  Code code = 2;
  string message = 3;
  AggregationValue data = 4;
  MessageType type = 5;
}

message Value {
  oneof value {
    // Single element
    double double = 1;
    sint64 sint64 = 2;
    bool bool = 3;
    string string = 4;
    bytes bytes = 5;
  }
}

message AggregationValue { map<string, Value> named_values = 1; }